src/VNCVideoCapture.cpp

```diff
@@ -0,0 +1,206 @@
+			rgba_buffer[index + 2] = client->frameBuffer[j + i + 2];
+		}
+	}
+
+
+	rtc::scoped_refptr<webrtc::I420Buffer> I420buffer = webrtc::I420Buffer::Create(client->width, client->height);
+	// RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "Transcoding frame ...";
+	const int conversionResult = libyuv::RAWToI420(rgba_buffer, client->width * 3,
+		(uint8_t*)I420buffer->DataY(), I420buffer->StrideY(),
+		(uint8_t*)I420buffer->DataU(), I420buffer->StrideU(),
+		(uint8_t*)I420buffer->DataV(), I420buffer->StrideV(),
+		client->width, client->height
+	);				
+	// RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "Finished Transcoding Frame!!! :D";
+					
+	if (conversionResult >= 0) {
+		webrtc::VideoFrame frame(I420buffer, 0, ts, webrtc::kVideoRotation_0);
+		// RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "Sending Frame!!! :D";
+		this->OnFrame(frame, frame.height(), frame.width());
+	} else {
+		RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "got back error decoding data" << conversionResult;
+		onError("Got back Error Handling Frame!!");
+	}
+
+	delete[] rgba_buffer;
+}
+
+VNCVideoCapturer::VNCVideoCapturer(const std::string u): client(rfbGetClient(8,3,4)), uri(u)
+{}
+
+bool VNCVideoCapturer::onStart() {
+	client->GetPassword = get_password;
+	client->FinishedFrameBufferUpdate = get_frame;
+
+	// remove vnc:// prefix from uri before passing along
+	std::string url = uri.substr(6, std::string::npos);
+	// if it contains password, remove it from the URL
+	auto index = url.find('@');
+	if (index != std::string::npos) {
+		url = url.substr(index + 1);
+	}
+	RTC_LOG(LERROR) << __PRETTY_FUNCTION__ << "Trying to initialize VNC with: " << url;
+
+	const char *args[] = { "VNCVideoCapture", url.c_str() };
+	int len = 2;
+	rfbClientSetClientData(client, (void *) "this", (void *) this);
+	if (!rfbInitClient(client, &len, (char **) args)) {
+		onError("Something went wrong in initializing RFB client ");
+		return false;
+	}
+	RTC_LOG(LERROR) << __PRETTY_FUNCTION__ << "Initialized VNC Successfully: " << url;
+	return true;
+}
+
+void VNCVideoCapturer::onError(char str[]) {
+	RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << str;
+	this->Stop();
+	// exit(EXIT_FAILURE);
+}
+
+VNCVideoCapturer::~VNCVideoCapturer()
+{
+	RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "~~~VNCVideoCapturer";
+}
+
+// overide rtc::Thread
+void VNCVideoCapturer::Run()
+{
+	RTC_LOG(LS_ERROR) << "VNCVideoCapturer Frame capture started ...";
+	if (!this->onStart()) {
+		return;
+	}
+	signal(SIGINT, signal_handler);
+	while (IsRunning()) {
+		// RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "Capturing Frame ...";
+		SendFramebufferUpdateRequest(client, 0, 0, client->width, client->height, FALSE);
+		if (WaitForMessage(client, 50) < 0) {
+			RTC_LOG(LS_ERROR) << "VNCVideoCapturer Frame capture timed out";
+			this->Stop();
+			break;
+		}
+		if (!HandleRFBServerMessage(client)) {
+			RTC_LOG(LS_ERROR) << "VNCVideoCapturer Frame capture cant handle message";
+			this->Stop();
+			break;
+		}
+	}
+}
```

Comment:
@ubergeek42 has a PR on beergoogles in flight on improving this logic - https://github.com/rainforestapp/beergoogles/pull/15 - prolly want to steal that.

Reply:
beautiful ❤️